ddl/ddl_api.go

```diff
@@ -2938,6 +2940,69 @@ func (d *ddl) ChangeColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.Al
 	return errors.Trace(err)
 }
 
+// RenameColumn renames an existing column.
+func (d *ddl) RenameColumn(ctx sessionctx.Context, ident ast.Ident, spec *ast.AlterTableSpec) error {
+	oldColName := spec.OldColumnName.Name
+	newColName := spec.NewColumnName.Name
+	if oldColName.L == newColName.L {
+		return nil
+	}
+
+	is := d.infoHandle.Get()
+	schema, ok := is.SchemaByName(ident.Schema)
+	if !ok {
+		return errors.Trace(infoschema.ErrDatabaseNotExists)
+	}
+
+	tbl, err := is.TableByName(ident.Schema, ident.Name)
+	if err != nil {
+		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ident.Schema, ident.Name))
+	}
+
+	oldCol := table.FindCol(tbl.VisibleCols(), oldColName.L)
+	if oldCol == nil {
+		return infoschema.ErrColumnNotExists.GenWithStackByArgs(oldColName, ident.Name)
+	}
+
+	colWithNewNameAlreadyExist := table.FindCol(tbl.Cols(), newColName.L) != nil
+	if colWithNewNameAlreadyExist {
+		return infoschema.ErrColumnExists.GenWithStackByArgs(newColName)
+	}
+
+	if fkInfo := getColumnForeignKeyInfo(oldColName.L, tbl.Meta().ForeignKeys); fkInfo != nil {
+		return errFKIncompatibleColumns.GenWithStackByArgs(oldColName, fkInfo.Name)
+	}
+
+	newCol := oldCol.Clone()
+	newCol.Name = newColName
+
+	// Check generated expression.
+	for _, col := range tbl.Cols() {
```

Comment:
We traverse `table.Cols()`/`table.VisiableCols()` 3 times in function `RenameColumn`, 
Can we traverse it only once?

Reply:
I think the performance difference is not significant between
```go
for i := 0; i < n; i+=1 {
    compute1()
    compute2()
    compute3()
}
```
and
```go
for i := 0; i < n; i+=1 {
    compute1()
}
for i := 0; i < n; i+=1 {
    compute2()
}
for i := 0; i < n; i+=1 {
    compute3()
}
```

Reply:
okay